src/shims/unix/linux/epoll.rs

```diff
@@ -415,43 +416,113 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        // is not close after the thread unblocks.
+        let weak_epfd = epfd.downgrade();
+
+        // We just need to know if the ready list is empty. The whole logic is wrapped inside a block
+        // so we don't need to manually drop epfd later.
+        let ready_list_empty;
+        {
+            let epoll_file_description = epfd
+                .downcast::<Epoll>()
+                .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_wait`"))?;
+            let binding = epoll_file_description.get_ready_list();
+            ready_list_empty = binding.borrow_mut().is_empty();
         }
 
-        let Some(epfd) = this.machine.fds.get(epfd) else {
-            return Ok(Scalar::from_i32(this.fd_not_found()?));
+        if ready_list_empty {
+            if timeout == 0 {
+                // Non-blocking with no notification returned.
+                this.write_scalar(Scalar::from_i32(0), &dest)?;
+                return Ok(());
+            } else {
+                // Blocking
+                let duration = Duration::from_millis(timeout.into());
+                this.block_thread(
+                    BlockReason::Epoll,
+                    Some((TimeoutClock::Monotonic, TimeoutAnchor::Relative, duration)),
+                    callback!(
+                        @capture<'tcx> {
+                            epfd_value: i32,
+                            weak_epfd: WeakFileDescriptionRef,
+                            dest: MPlaceTy<'tcx>,
+                            event: MPlaceTy<'tcx>,
+                        }
+                        @unblock = |this| {
+                            this.blocking_epoll_callback(epfd_value, weak_epfd, &dest, &event)?;
+                            Ok(())
+                        }
+                        @timeout = |this| {
+                            // No notification after blocking timeout.
+                            this.write_scalar(Scalar::from_i32(0), &dest)?;
+                            Ok(())
+                        }
+                    ),
+                );
+            }
+        } else {
+            // Immediately return epoll notification without blocking.
+            this.blocking_epoll_callback(epfd_value, weak_epfd, &dest, &event)?;
+            return Ok(());
+        }
+
+        Ok(())
+    }
+
+    /// Callback function after epoll_wait unblocks
+    fn blocking_epoll_callback(
+        &mut self,
+        epfd_value: i32,
+        weak_epfd: WeakFileDescriptionRef,
+        dest: &MPlaceTy<'tcx>,
+        event: &MPlaceTy<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let Some(epfd) = weak_epfd.upgrade() else {
+            throw_unsup_format!("epoll FD {epfd_value} is closed while blocking.")
         };
+
+        //TODO: What if, there is another thread that make changes to the global fd table here?
+        // For example close the current epfd, then insert a new one inside?
```

Comment:
Since we're using a file description ref, that should not cause any issues now, right?

Reply:
Yes, this shouldn't be a problem. It will only be a problem if we use the file descriptor value to retrieve the file description ref again.